atc/exec/build/repository.go

```diff
@@ -68,3 +79,13 @@ func (repo *Repository) AsMap() map[ArtifactName]runtime.Artifact {
 
 	return result
 }
+
+func (repo *Repository) NewLocalScope() *Repository {
+	child := NewRepository()
+	child.parent = repo
+	return child
+}
+
+func (repo *Repository) Parent() *Repository {
```

Comment:
Yeah I'm iffy on it too - I did the same for `(exec.RunState).Parent`). It's a bit of a trade-off. Making it a tested API contract allows the tests for things which use `NewLocalScope()` to just call `.Parent()` to assert that they're nested. The alternative I see is to behaviorally test the artifact scoping mechanics in each place, but that seemed a bit too repetitive and brought the tests beyond their core competency (IMO `NewLocalScope` should be cheap to use and test).

So at this point it's an API contract covered in tests, and right now it's only there to make other tests a bit easier to write. It's not used in production, but it also shouldn't require any maintenance. It does make it easier to for someone to use it in the future, but it's also so trivial and that it would probably just be exported/added if it's not already there, so 🤷‍♂️ I'm OK with the trade-offs myself. Happy to chat more tho.

Reply:
Hehe, you did add behavioural tests for `AsMap` and `ArtfactFor` in `repository_test.go` and it looks good :)
So, I feel like it should be safe to drop `Parent`. I agree it is trivial to add in but the action would have to be intentional. But overall, this is minor.